src/field/field255.rs

```diff
@@ -0,0 +1,483 @@
+    }
+}
+
+impl PartialEq for Field255 {
+    fn eq(&self, rhs: &Self) -> bool {
+        // The internal representation used by fiat-crypto is not 1-1 with the field, so it is
+        // necessary to compare field elements via their canonical encodings.
+
+        let mut self_encoded = [0; 32];
+        fiat_25519_to_bytes(&mut self_encoded, &self.0);
+        let mut rhs_encoded = [0; 32];
+        fiat_25519_to_bytes(&mut rhs_encoded, &rhs.0);
+
+        let mut equal = Choice::from(1);
+        for (lhs_byte, rhs_byte) in self_encoded.iter().zip(rhs_encoded.iter()) {
+            equal &= lhs_byte.ct_eq(rhs_byte)
+        }
+        equal.into()
+    }
+}
+
+impl Eq for Field255 {}
+
+impl Add for Field255 {
+    type Output = Field255;
+
+    fn add(self, rhs: Self) -> Field255 {
+        let mut loose_output = [0; 5];
+        fiat_25519_add(&mut loose_output, &self.0, &rhs.0);
+        let mut output = [0; 5];
+        fiat_25519_carry(&mut output, &loose_output);
+        Field255(output)
+    }
+}
+
+impl AddAssign for Field255 {
+    fn add_assign(&mut self, rhs: Self) {
+        let mut loose_output = [0; 5];
+        fiat_25519_add(&mut loose_output, &self.0, &rhs.0);
+        fiat_25519_carry(&mut self.0, &loose_output);
+    }
+}
+
+impl Sub for Field255 {
+    type Output = Field255;
+
+    fn sub(self, rhs: Self) -> Field255 {
+        let mut loose_output = [0; 5];
+        fiat_25519_sub(&mut loose_output, &self.0, &rhs.0);
+        let mut output = [0; 5];
+        fiat_25519_carry(&mut output, &loose_output);
+        Field255(output)
+    }
+}
+
+impl SubAssign for Field255 {
+    fn sub_assign(&mut self, rhs: Self) {
+        let mut loose_output = [0; 5];
+        fiat_25519_sub(&mut loose_output, &self.0, &rhs.0);
+        fiat_25519_carry(&mut self.0, &loose_output);
+    }
+}
+
+impl Mul for Field255 {
+    type Output = Field255;
+
+    fn mul(self, rhs: Self) -> Field255 {
+        let mut output = [0; 5];
+        fiat_25519_carry_mul(&mut output, &self.0, &rhs.0);
+        Field255(output)
+    }
+}
+
+impl MulAssign for Field255 {
+    fn mul_assign(&mut self, rhs: Self) {
+        let self_copy = self.0;
+        fiat_25519_carry_mul(&mut self.0, &self_copy, &rhs.0);
+    }
+}
+
+impl Div for Field255 {
+    type Output = Field255;
+
+    fn div(self, _rhs: Self) -> Self::Output {
+        unimplemented!("Div is not implemented for Field255 because it's not needed yet")
+    }
+}
+
+impl DivAssign for Field255 {
+    fn div_assign(&mut self, _rhs: Self) {
+        unimplemented!("DivAssign is not implemented for Field255 because it's not needed yet")
+    }
+}
+
+impl Neg for Field255 {
+    type Output = Field255;
+
+    fn neg(self) -> Field255 {
+        let mut loose_output = [0; 5];
+        fiat_25519_opp(&mut loose_output, &self.0);
+        let mut output = [0; 5];
+        fiat_25519_carry(&mut output, &loose_output);
+        Field255(output)
+    }
+}
+
+impl TryFrom<usize> for Field255 {
```

Comment:
`TryFrom<usize>` feels sketchy given that `Field255` values are much, much bigger than any `usize` type. Is this some kind of scheme to "compress" sufficiently small values into 8 bytes? Why do we need this?

Reply:
`TryFrom<usize>` is a (newly added) trait bound on `FieldElement`. We need some sort of bound like this so that we can construct field elements from integers when we don't have access to the `FieldElementWithInteger` trait. Poplar1 will use this by casting control bits to `usize`, then converting them to field elements, and using that resulting element to compute IDPF output values without branching on the control bit.

Reply:
What concerns me here is that carrying this `TryFrom` impl on `Field255` implies that one can generally convert a `usize` to a `Field255`, which I don't think is true and I worry it could lead to bugs if misused by clients. It sounds like we have a specific use case of wanting to transform "control bits" into a `Field255`, rather than the general requirement of transforming `usize` into `Field255`. So assuming that some kind of `struct Poplar1ControlBits` appears, could we instead do `TryFrom<Poplar1ControlBits> for Field255`?

Reply:
I would go with `u64` instead of `usize`, for consistency. (See comment above.)

Reply:
Any `usize` can be successfully converted to `Field255` if `usize` is 32 or 64 bits. For CHERI-like platforms where `usize` is 128 or 129 bits, the same holds. You'd need to have `usize` be 256 bits or larger to run into issues. (At 256 bits, `try_from_bytes` may return `FieldError::ModulusOverflow`, and beyond that, a slicing operation would panic)

Implementing `TryFrom<usize> for Field255` was motivated by [this comment thread](https://github.com/divviup/libprio-rs/pull/408#discussion_r1065237379). I'd like to avoid implementing conversions from more than one primitive integer type, as this would impede type inference of integer literals. Implementing `TryFrom<subtle::Choice> for Field255`/`TryFrom<sublte::Choice> for Field64` would be sufficient for the IDPF use case, but `TryFrom<usize> for Field...` is more generally applicable, and can simplify some Prio3 code that currently bounces throgh two conversions.

Reply:
All right, I'm willing to defer to your consensus since I think you both have a better understanding of how this will fit into a Poplar1 implementation.